resources/virt.go

```diff
@@ -0,0 +1,521 @@
+package resources
+
+import (
+	"encoding/gob"
+	"log"
+	"fmt"
+	"strconv"
+
+	"github.com/purpleidea/mgmt/event"
+	"github.com/rgbkrk/libvirt-go"
+	"math/rand"
+	"time"
+	"bytes"
+)
+
+func init() {
+	gob.Register(&VirtRes{})
+}
+
+type VirtRes struct {
+	BaseRes `yaml:",inline"`
+	Cpus int `yaml:"cpus"`
+	Ram int `yaml:"ram"`
+	State string `yaml:"state"` //running, paused, shutoff
+	Transient bool `yaml:"transient"`
+	Boot []string `yaml:"boot"` // boot order. values: fd, hd, cdrom, network
+	Disk []diskDevice `yaml:"disk"`
+	CDRom []cdRomDevice `yaml:"cdrom"`
+	Network []networkDevice `yaml:"network"`
+	Filesystem []filesystemDevice `yaml:"filesystem"`
+	conn libvirt.VirConnection
+}
+
+type VirtUUID struct {
+	BaseUUID
+}
+
+type virtDevice interface {
+	GetXML(idx int) string
+}
+
+type diskDevice struct {
+	Source string `yaml:"source"`
+	Type string `yaml:"type"`
+}
+
+type cdRomDevice struct {
+	Source string `yaml:"source"`
+	Type string `yaml:"type"`
+}
+
+type networkDevice struct {
+	Name string `yaml:"name"`
+	MAC string `yaml:"mac"`
+}
+
+type filesystemDevice struct {
+	Access string `yaml:"access"`
+	Source string `yaml:"source"`
+	Target string `yaml:"target"`
+	ReadOnly bool `yaml:"read_only"`
+}
+
+var (
+	libvirtInitialized bool = false
+)
+
+func NewVirtRes(name string, cpus, ram int, state string, transient bool) *VirtRes {
+	obj := &VirtRes{
+		BaseRes: BaseRes{
+			Name: name,
+		},
+		Cpus: cpus,
+		Ram: ram,
+		State: state,
+		Transient: transient,
+	}
+	obj.Init()
+	return obj
+}
+
+func (obj *VirtRes) Init() error {
+	obj.BaseRes.kind = "Virt"
+
+	if !libvirtInitialized {
+		libvirt.EventRegisterDefaultImpl()
+		// run libvirt event loop
+		go func() {
```

Comment:
How does this work? Do we ever have to shut this loop down somehow? If so, and if this requires an API change to the resource API LMK how we can help. Getting your feedback on if we designed this correctly is very valuable to us!